src/modal/modal.js

```diff
@@ -0,0 +1,336 @@
+      return;
+    }
+
+    if (!this.props.closeable) {
+      return;
+    }
+
+    this.close(CLOSE_SOURCE.escape);
+  };
+
+  onBackdropClick = () => {
+    if (!this.props.closeable) {
+      return;
+    }
+    this.close(CLOSE_SOURCE.backdrop);
+  };
+
+  onCloseClick = () => {
+    this.close(CLOSE_SOURCE.closeButton);
+  };
+
+  clearTimers() {
+    if (this.animateOutTimer) {
+      clearTimeout(this.animateOutTimer);
+    }
+    if (this.animateStartTimer) {
+      // eslint-disable-next-line cup/no-undef
+      cancelAnimationFrame(this.animateStartTimer);
+    }
+  }
+
+  open() {
+    // Store last focused item (we'll return focus after closing for a11y)
+    this.captureLastFocus();
+    this.autoFocus();
+
+    // Sometimes scroll starts past zero, possibly due to animation
+    // Reset scroll to 0 (other libraries do this as well)
+    const rootRef = this.getRef('Root').current;
+    if (rootRef) {
+      rootRef.scrollTop = 0;
+    }
+
+    // Clear any existing timers (like previous animateOutTimer)
+    this.clearTimers();
+    this.addDomEvents();
+
+    // eslint-disable-next-line cup/no-undef
+    this.animateStartTimer = requestAnimationFrame(() => {
+      this.setState({isVisible: true});
+    });
+  }
+
+  close(source?: CloseSourceT) {
+    // If there's no source, it just means the isOpen prop changed. No need to call onClose.
+    if (this.props.onClose && source) {
+      this.props.onClose(source);
+    }
+    this.removeDomEvents();
+    this.animateOutTimer = setTimeout(this.animateOutComplete, 500);
+    this.restoreLastFocus();
+  }
+
+  captureLastFocus = () => {
+    this.lastFocus = ownerDocument(this.getMountNode()).activeElement;
+  };
+
+  restoreLastFocus = () => {
+    if (this.lastFocus) {
+      // Not all elements in IE11 can focus
+      if (this.lastFocus.focus) {
+        this.lastFocus.focus();
+      }
+      this.lastFocus = null;
+    }
+  };
+
+  autoFocus = () => {
+    if (!this.props.autofocus) {
+      return;
+    }
+    const dialog = this.getRef('Dialog').current;
+    if (!dialog) {
+      return;
+    }
+    dialog.focus();
+  };
+
+  animateOutComplete = () => {
+    this.setState({
+      isVisible: false,
+    });
+  };
+
+  getSharedProps(): $Diff<SharedStylePropsArgT, {children: React.Node}> {
+    const {animate, isOpen, size, role, closeable} = this.props;
+    return {
+      $animate: animate,
+      $isVisible: this.state.isVisible,
+      $isOpen: Boolean(isOpen),
+      $size: size,
+      $role: role,
+      $closeable: Boolean(closeable),
+    };
+  }
+
+  getMountNode(): HTMLElement {
+    const {mountNode} = this.props;
+    if (mountNode) {
+      return mountNode;
+    }
+    // Flow thinks body could be null (cast through any)
+    // eslint-disable-next-line flowtype/no-weak-types
+    return ((document.body: any): HTMLBodyElement);
+  }
+
+  getChildren() {
+    const {children} = this.props;
+    return typeof children === 'function' ? children() : children;
```

Comment:
If a modal is hidden, it seems a tad wasteful to still pass all the children–even if its essentially just a bunch of React.createElement calls. I was thinking optionally supporting children as function was a cheap addition and allows people to lazily render children only when the modal is open. That said, I don't feel super strongly here, if you think it could lead to confusion we can remove it.